irc/histserv.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2020 Shivaram Lingamneni <slingamn@cs.stanford.edu>
+// released under the MIT license
+
+package irc
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/oragono/oragono/irc/history"
+)
+
+const (
+	histservHelp = `HistServ provides commands related to history.`
+	histServMask = "HistServ!HistServ@localhost"
+)
+
+func histservEnabled(config *Config) bool {
+	return config.History.Enabled
+}
+
+func historyComplianceEnabled(config *Config) bool {
+	return config.History.Enabled && config.History.Persistent.Enabled && config.History.Retention.EnableAccountIndexing
+}
+
+var (
+	histservCommands = map[string]*serviceCommand{
+		"forget": {
+			handler: histservForgetHandler,
+			help: `Syntax: $bFORGET <account>$b
+
+FORGET deletes all history messages sent by an account.`,
+			helpShort: `$bFORGET$b deletes all history messages sent by an account.`,
+			capabs:    []string{"history"},
+			enabled:   histservEnabled,
+			minParams: 1,
+			maxParams: 1,
+		},
+		"delete": {
+			handler: histservDeleteHandler,
+			help: `Syntax: $bDELETE [target] <msgid>$b
+
+DELETE deletes an individual message by its msgid. The target is a channel
+name or nickname; depending on the history implementation, this may or may not
+be necessary to locate the message.`,
+			helpShort: `$bDELETE$b deletes an individual message by its msgid.`,
+			enabled:   histservEnabled,
+			minParams: 1,
+			maxParams: 2,
+		},
+		"export": {
+			handler: histservExportHandler,
+			help: `Syntax: $bEXPORT <account>$b
+
+EXPORT exports all messages sent by an account as JSON. This can be used
+for regulatory compliance, e.g., article 15 of the GDPR.`,
```

Comment:
Let's not mention GDPR unless we have strict advice about the applicability of our methods to it? Also probably cleaner to just mention what it does and leave it there.

Reply:
I left in the mention of "compliance" because otherwise this sounds kind of creepy.

Now that I think about it, it's even creepier if I say "compliance" without mentioning the GDPR --- it sounds like this is for turning over data to a three-letter agency :-\

I should think more about how to document this. It's not entirely clear to me how to advertise this feature. I think being able to provide "strict advice" will be impossible (we're not lawyers, and even if we were, we're not in an attorney-client relationship with the users), which shouldn't mean that we can't provide some guidance.